Decompile and clean up some effect files

Roman971 opened the pull request:
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---
<!-- Leave the text above intact. Add additional comments below. -->
This includes most files from the "static" effect system (the other one besides soft sprites), along with `z_eff_ss_dead.c` which just contains some utility graphic functions. These files haven't really changed a lot from OoT, tho the notable exception is `z_eff_tire_mark.c` which is a new type of effect added in MM.
I also went through the related structs and engine code to clean things up a bit and include some more recent documentation, notably a new enum for effect types and some defines.

Note: I added a `(void)0;` in the `CLOSE_DISPS` macro to allow it to be used directly after a goto label, which was needed in `EffectSpark_Draw`

EllipticEllipsis commented on include/macros.h:
```diff
@@ -96,6 +96,7 @@ extern GraphicsContext* __gfxCtx;
         s32 __dispPad
 
 #define CLOSE_DISPS(gfxCtx) \
+    (void)0;                \
```
Would it be better to to this in the one place that it's needed than change the macro?

Roman971 replied:
The idea is to make it so you can use the macro without context, and to have the same behavior as the debug version, since the debug version has a function call which acts as the valid statement before the bracket.
Also it's probably going to be needed in more places because using a goto to CLOSE_DISPS is the "right" way to do an early return in a draw function. So I think we might as well add it to the macro now and avoid more headache in the future.

EllipticEllipsis replied:
I see, makes sense.

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+
+    if (this != NULL) {
+        numElements = this->numElements;
+        if (numElements >= 16) {
```
```suggestion
        if (numElements >= ARRAY_COUNT(this->elements)) {
```

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+
+    this->numElements = 0;
+
+    for (i = 0; i < 16; i++) {
```
```suggestion
    for (i = 0; i < ARRAY_COUNT(this->elements); i++) {
```

engineer124 commented on include/z64effect.h:
```diff
@@ -82,13 +82,15 @@ typedef struct {
-    /* 0x194 */ UNK_TYPE1 pad194[0xC];
-} EffBlureInit1; // size = 0x1A0
+    /* 0x000 */ char unk_00[0x184];
+    /* 0x184 */ u8 p1StartColor[4];
+    /* 0x188 */ u8 p2StartColor[4];
+    /* 0x18C */ u8 p1EndColor[4];
+    /* 0x190 */ u8 p2EndColor[4];
```
So these have to be arrays instead of `Color_RGBA8`?

Roman971 replied:
These had to be arrays in OoT because they were set with for loops in some actors (note that only the "init" structs have them as arrays), so I left it this way for now, but I guess we'll have to see if it remains needed in MM

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+
+    while (true) {
+        if (this->elements[0].state == 0) {
+            for (i = 0; i < 15; i++) {
```
```suggestion
            for (i = 0; i < ARRAY_COUNT(this->elements) - 1; i++) {
```

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    }
+
+    if (this->elemDuration < this->elements[0].timer) {
+        for (i = 0; i < 15; i++) {
```
```suggestion
        for (i = 0; i < ARRAY_COUNT(this->elements) - 1; i++) {
```

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    f32 sp2C;
+
+    if (((elem - 1)->state == 0) || ((elem + 1)->state == 0)) {
+        elem->flags &= ~3;
```
```suggestion
        elem->flags &= ~(2 | 1);
```
My personal preference for bitwise flags is to write them out explicitly so that it's easier to see which bits are turned off, plus it'll be easier to document flag enums down the road (applies to a few places)

Roman971 replied:
Is this the preference in MM? In OoT we've generally avoided doing this, because all these flags would eventually have defines anyway, and it's slightly easier to automate splitting everything into defines if they are kept as the combined value.
But if MM is doing things differently, I don't mind making the change here

engineer124 replied:
I would say there isn't an MM preference, this has been more of a personal preference. Does it make it easier for scripts to split? I've always found it easier to read and document with it split, but I'm open to changing it if it's more convenient for tools

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+            Math3D_AngleBetweenVectors(&sp4C, &sp40, &sp2C)) {
+            elem->flags &= ~3;
+            elem->flags |= 0;
+        } else if ((sp34 <= -0.5f) || (sp30 <= -0.5f) || (sp2C <= 0.7071f)) {
```
```suggestion
        } else if ((sp34 <= -0.5f) || (sp30 <= -0.5f) || (sp2C <= 0.7071f)) { // sp2C <= sqrt(2) / 2
```
I wonder if this comment is necessary or if there's a cleaner way to represent this

Roman971 replied:
I feel like it would a bit weird to have a comment specifically for this part of the `||`, but if there is a cleaner way to write the float, then we should definitely use it. I don't have any idea myself tho

engineer124 replied:
Yeah that's fine, I was just throwing an idea out there. I can't think of any good idea myself

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+
+void EffectBlure_DrawElemHermiteInterpolation(EffectBlure* this, EffectBlureElement* elem, s32 index,
+                                              GraphicsContext* gfxCtx) {
+    static Vtx_t baseVtx = VTX_T(0, 0, 0, 0, 0, 0xFF, 0xFF, 0xFF, 0xFF);
```
```suggestion
    static Vtx_t baseVtx = VTX_T(0, 0, 0, 0, 0, 255, 255, 255, 255);
```
To be consistent with the function above

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    Math_Vec3s_ToVec3f(&sp18C, &sp1EC);
+    Math_Vec3s_ToVec3f(&sp180, &sp1E4);
+
+    if ((elem->flags & 3) == 2) {
```
```suggestion
    if ((elem->flags & (2 | 1)) == 2) {
```
imo this is easier to read since it's clearer that it's checking to make sure both bit 2 is on and bit 1 is off, but I could understand if the brackets look a little funny

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    Math_Vec3f_Scale(&sp1B4, 0.5f);
+    Math_Vec3f_Scale(&sp1A8, 0.5f);
+
+    if (((elem + 1)->flags & 3) == 2) {
```
```suggestion
    if (((elem + 1)->flags & (2 | 1)) == 2) {
```

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+            s32 j2 = 2 * i + 1;
+            Vec3f spE0;
+            f32 temp_f28 = i / 7.0f;                               // t
+            f32 temp_f0 = temp_f28 * temp_f28;                     // t^2
```
```suggestion
            f32 temp_f0 = SQ(temp_f28);                     // t^2
```
May need to reformat to ensure the `// t^2` comment lines up

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+        return;
+    }
+
+    this->elements[0].flags &= ~3;
+    this->elements[0].flags |= 2;
```
So this turns off bits 1 & 2, only to turn on bit 2 again?

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+                sp1A4.z = ((f32)vtx[4 * i + 2].v.ob[2] + (f32)vtx[4 * i + 3].v.ob[2]) * 0.5f;
+
+                Math_Vec3f_Diff(&sp1A4, &sp1B0, &sp198);
+                scale = sqrtf(SQ(sp198.x) + SQ(sp198.y) + SQ(sp198.z));
```
```suggestion
                scale = sqrtf(SQXYZ(sp198));
```

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+                    Math_Vec3f_Scale(&sp198, scale);
+
+                    SkinMatrix_SetTranslate(&sp154, sp1B0.x, sp1B0.y, sp1B0.z);
+                    SkinMatrix_SetRotateAroundVec(&sp114, 0x3FFF, sp198.x, sp198.y, sp198.z);
```
No suggestion here, but an observation that for 90 degree and 180 degree angles, OoT commonly has those angles defined as `0x3FFF` and `0x7FFF` respectively, while MM has been very consistent in replacing those with angles `0x4000` and `0x8000` respectively. This seems like a rare case in MM where `0x3FFF` still remains. Figured that was worth commenting on

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    CLOSE_DISPS(gfxCtx);
+}
+
+Vtx_t D_801AE26C[] = {
+    VTX_T(0, 0, 0, 0, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX_T(0, 0, 0, 0, 0, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX_T(0, 0, 0, 2048, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX_T(0, 0, 0, 2048, 0, 0xFF, 0xFF, 0xFF, 0xFF),
+};
+
+Vtx_t D_801AE2AC[] = {
+    VTX_T(0, 0, 0, 2048, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX_T(0, 0, 0, 2048, 0, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX_T(0, 0, 0, 2048, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX_T(0, 0, 0, 2048, 0, 0xFF, 0xFF, 0xFF, 0xFF),
+};
```
```suggestion
static Vtx_t D_801AE26C[] = {
    VTX_T(0, 0, 0, 0, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
    VTX_T(0, 0, 0, 0, 0, 0xFF, 0xFF, 0xFF, 0xFF),
    VTX_T(0, 0, 0, 2048, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
    VTX_T(0, 0, 0, 2048, 0, 0xFF, 0xFF, 0xFF, 0xFF),
};

static Vtx_t D_801AE2AC[] = {
    VTX_T(0, 0, 0, 2048, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
    VTX_T(0, 0, 0, 2048, 0, 0xFF, 0xFF, 0xFF, 0xFF),
    VTX_T(0, 0, 0, 2048, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
    VTX_T(0, 0, 0, 2048, 0, 0xFF, 0xFF, 0xFF, 0xFF),
};
```
This is a very small set of Vtx data so it's less of a big deal. But in general, MM has taken a more hardline approach to making sure all possible asset data, no matter how small or trivial, be extracted if possible. I think my preference would be to have this extracted, but I'll leave this up to you given the context

AngheloAlf replied:
Yeah. We are requiring every Vtx and texture to be extracted.
This is code tho, so the XML to extract them will be weird. If you need help, we could try to figure out a way to do it in discord.

Roman971 replied:
Imho basic vertices like these should be treated as an exception, especially in this case since they don't correspond to anything on their own and are just used as "base" vertices which are then set through code to contain the data needed, if that makes sense (see `EffectBlure_DrawSimple` for instance).

Roman971 replied:
That being said, we can try to extract them if it's really necessary, but we would first need ZAPD to handle `VTX_T` somehow, and any vertices that are in-function static will be difficult to extract to say the least

AngheloAlf replied:
In-function static Vtx shouldn't be a problem tbh, since ZAPD can extract Vtx to standalone `.vtx.inc` files. It would look like something like this:
```c
void function() {
  static Vtx someVtx[] = {
    #include "path/somevtx.vtx.inc"
  }

  // ...
}
```
Currently this isn't 100% reliable because of a few ZAPD bugs which prevent the extraction to standalone files. Those bugs should be fixed in PR https://github.com/zeldaret/ZAPD/pull/151.

Second, what's the difference between a VTX and a VTX_T? The macro looks similar enough. From a extraction pov looks like ZAPD would only need to replace `VTX` to `VTX_T`.

Finally, I can see why vertices like this could be treated as an exception, but I think this needs a further discussion.

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+                                vtx[j + 1].v.ob[0] = func_800B09D0(elem->p2.x, elem->p1.x, ratio);
+                                vtx[j + 1].v.ob[1] = func_800B09D0(elem->p2.y, elem->p1.y, ratio);
+                                vtx[j + 1].v.ob[2] = func_800B09D0(elem->p2.z, elem->p1.z, ratio);
+                                ratio = ratio + ratio;
```
```suggestion
                                ratio += ratio;
```
Is that clearer?

Roman971 replied:
I ended up changing it to *= 2.0f, which is the clearest option imo, and it's consistent with the *= 0.5f a few lines above.

engineer124 commented on src/code/z_eff_shield_particle.c:
```diff
@@ -1,11 +1,211 @@
+#include "vt.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eff_shield_particle/EffectShieldParticle_Init.s")
+static Vtx sVertices[] = {
+    VTX(-32, -32, 0, 0, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX(32, 32, 0, 1024, 0, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX(-32, 32, 0, 0, 0, 0xFF, 0xFF, 0xFF, 0xFF),
+    VTX(32, -32, 0, 1024, 1024, 0xFF, 0xFF, 0xFF, 0xFF),
+};
```
Same comment above about extraction

engineer124 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    EffectBlure_SetupSimpleAlt,
+};
+
+s32 D_801AE268 = 0; // unused
```
```suggestion
static s32 D_801AE268 = 0; // unused
```

Roman971 replied:
I didn't include the `static` for these unnamed symbols so they can remain in the map file and wouldn't conflict with anything, but I can add `static` if it's preferred

engineer124 replied:
I always assumed static prevented conflicts with other files by keeping it local, but that could very well be a misunderstanding on my part. I'm not sure what the standard is in OoT for unused variables set to zero, but could you just call it `sIsZero` and be done with it?

engineer124 commented on src/code/z_eff_shield_particle.c:
```diff
@@ -1,11 +1,211 @@
+        envColor->b = this->envColorStart.b;
+        envColor->a = this->envColorStart.a;
+    } else if (this->timer < halfDuration) {
+        f32 ratio = this->timer / (f32)halfDuration;
+        primColor->r = this->primColorStart.r + (this->primColorMid.r - this->primColorStart.r) * ratio;
```
```suggestion
        f32 ratio = this->timer / (f32)halfDuration;
        
        primColor->r = this->primColorStart.r + (this->primColorMid.r - this->primColorStart.r) * ratio;
```

engineer124 commented on src/code/z_eff_shield_particle.c:
```diff
@@ -1,11 +1,211 @@
+        envColor->b = this->envColorStart.b + (this->envColorMid.b - this->envColorStart.b) * ratio;
+        envColor->a = this->envColorStart.a + (this->envColorMid.a - this->envColorStart.a) * ratio;
+    } else {
+        f32 ratio = (this->timer - halfDuration) / (f32)halfDuration;
+        primColor->r = this->primColorMid.r + (this->primColorEnd.r - this->primColorMid.r) * ratio;
```
```suggestion
        f32 ratio = (this->timer - halfDuration) / (f32)halfDuration;
        
        primColor->r = this->primColorMid.r + (this->primColorEnd.r - this->primColorMid.r) * ratio;
```

engineer124 commented on src/code/z_eff_shield_particle.c:
```diff
@@ -1,11 +1,211 @@
+        envColor->a = this->envColorStart.a;
+    } else if (this->timer < halfDuration) {
+        f32 ratio = this->timer / (f32)halfDuration;
+        primColor->r = this->primColorStart.r + (this->primColorMid.r - this->primColorStart.r) * ratio;
+        primColor->g = this->primColorStart.g + (this->primColorMid.g - this->primColorStart.g) * ratio;
+        primColor->b = this->primColorStart.b + (this->primColorMid.b - this->primColorStart.b) * ratio;
+        primColor->a = this->primColorStart.a + (this->primColorMid.a - this->primColorStart.a) * ratio;
+        envColor->r = this->envColorStart.r + (this->envColorMid.r - this->envColorStart.r) * ratio;
+        envColor->g = this->envColorStart.g + (this->envColorMid.g - this->envColorStart.g) * ratio;
+        envColor->b = this->envColorStart.b + (this->envColorMid.b - this->envColorStart.b) * ratio;
+        envColor->a = this->envColorStart.a + (this->envColorMid.a - this->envColorStart.a) * ratio;
```
```suggestion
        primColor->r = F32_LERPIMP(this->primColorStart.r, this->primColorMid.r, ratio);
        primColor->g = F32_LERPIMP(this->primColorStart.g, this->primColorMid.g, ratio);
        primColor->b = F32_LERPIMP(this->primColorStart.b, this->primColorMid.b, ratio);
        primColor->a = F32_LERPIMP(this->primColorStart.a, this->primColorMid.a, ratio);
        envColor->r = F32_LERPIMP(this->envColorStart.r, this->envColorMid.r, ratio);
        envColor->g = F32_LERPIMP(this->envColorStart.g, this->envColorMid.g, ratio);
        envColor->b = F32_LERPIMP(this->envColorStart.b, this->envColorMid.b, ratio);
        envColor->a = F32_LERPIMP(this->envColorStart.a, this->envColorMid.a, ratio);
```

Roman971 replied:
I made a new `LERPIMP` which isn't restricted to floats like you suggested, but I also left the `F32_LERPIMP` for float purposes and added f32 casts to it so it forces floats to be used and avoids casts in the case of `z_eff_spark.c`

engineer124 commented on src/code/z_eff_shield_particle.c:
```diff
@@ -1,11 +1,211 @@
+        envColor->a = this->envColorStart.a + (this->envColorMid.a - this->envColorStart.a) * ratio;
+    } else {
+        f32 ratio = (this->timer - halfDuration) / (f32)halfDuration;
+        primColor->r = this->primColorMid.r + (this->primColorEnd.r - this->primColorMid.r) * ratio;
+        primColor->g = this->primColorMid.g + (this->primColorEnd.g - this->primColorMid.g) * ratio;
+        primColor->b = this->primColorMid.b + (this->primColorEnd.b - this->primColorMid.b) * ratio;
+        primColor->a = this->primColorMid.a + (this->primColorEnd.a - this->primColorMid.a) * ratio;
+        envColor->r = this->envColorMid.r + (this->envColorEnd.r - this->envColorMid.r) * ratio;
+        envColor->g = this->envColorMid.g + (this->envColorEnd.g - this->envColorMid.g) * ratio;
+        envColor->b = this->envColorMid.b + (this->envColorEnd.b - this->envColorMid.b) * ratio;
+        envColor->a = this->envColorMid.a + (this->envColorEnd.a - this->envColorMid.a) * ratio;
```
```suggestion
        primColor->r = F32_LERPIMP(this->primColorMid.r, this->primColorEnd.r, ratio);
        primColor->g = F32_LERPIMP(this->primColorMid.g, this->primColorEnd.g, ratio);
        primColor->b = F32_LERPIMP(this->primColorMid.b, this->primColorEnd.b, ratio);
        primColor->a = F32_LERPIMP(this->primColorMid.a, this->primColorEnd.a, ratio);
        envColor->r = F32_LERPIMP(this->envColorMid.r, this->envColorEnd.r, ratio);
        envColor->g = F32_LERPIMP(this->envColorMid.g, this->envColorEnd.g, ratio);
        envColor->b = F32_LERPIMP(this->envColorMid.b, this->envColorEnd.b, ratio);
        envColor->a = F32_LERPIMP(this->envColorMid.a, this->envColorEnd.a, ratio);
```

engineer124 commented on src/code/z_eff_shield_particle.c:
```diff
@@ -1,11 +1,211 @@
+            MtxF sp104;
+            MtxF spC4;
+            MtxF sp84;
+            f32 temp1 = ((elem->endX + elem->startX) * 0.5f);
+            f32 temp2 = elem->endX - elem->startX;
+            f32 temp3 = ((temp2 * (1.0f / 64.0f)) / 0.02f);
```
```suggestion
            f32 temp1 = (elem->endX + elem->startX) * 0.5f;
            f32 temp2 = elem->endX - elem->startX;
            f32 temp3 = (temp2 * (1.0f / 64.0f)) / 0.02f;
```

engineer124 commented on src/code/z_eff_spark.c:
```diff
@@ -1,9 +1,265 @@
+            elem->velocity.y = Rand_ZeroOne() - 0.5f;
+            elem->velocity.z = Rand_ZeroOne() - 0.5f;
+
+            velocityNorm = sqrtf(SQ(elem->velocity.x) + SQ(elem->velocity.y) + SQ(elem->velocity.z));
```
```suggestion
            velocityNorm = sqrtf(SQXYZ(elem->velocity));
```

engineer124 commented on src/code/z_eff_spark.c:
```diff
@@ -1,9 +1,265 @@
+
+            velocityNorm = sqrtf(SQ(elem->velocity.x) + SQ(elem->velocity.y) + SQ(elem->velocity.z));
+
+            if (!(fabsf(velocityNorm) < 0.008f)) {
```
```suggestion
            if (!IS_ZERO(velocityNorm)) {
```
I'm not sure I like this macro since zero checks are all over the place but the value isn't always `0.008f`, but I suppose it does work here

Roman971 replied:
I'm not sure it makes sense to use this macro outside collision stuff since `0.008f` doesn't necessarily work as an epsilon depending on the floats being used, so abstracting it away could lead to issues in some contexts, and it can't be used consistently like you mentioned

engineer124 commented on src/code/z_eff_spark.c:
```diff
@@ -1,9 +1,265 @@
+        j = 0;
+
+        ratio = (f32)this->timer / (f32)this->duration;
+        sp1D3 = this->colorStart[0].r + ((f32)this->colorEnd[0].r - (f32)this->colorStart[0].r) * ratio;
+        sp1D2 = this->colorStart[0].g + ((f32)this->colorEnd[0].g - (f32)this->colorStart[0].g) * ratio;
+        sp1D1 = this->colorStart[0].b + ((f32)this->colorEnd[0].b - (f32)this->colorStart[0].b) * ratio;
+        sp1D0 = this->colorStart[0].a + ((f32)this->colorEnd[0].a - (f32)this->colorStart[0].a) * ratio;
+        sp1CF = this->colorStart[1].r + ((f32)this->colorEnd[1].r - (f32)this->colorStart[1].r) * ratio;
+        sp1CE = this->colorStart[1].g + ((f32)this->colorEnd[1].g - (f32)this->colorStart[1].g) * ratio;
+        sp1CD = this->colorStart[1].b + ((f32)this->colorEnd[1].b - (f32)this->colorStart[1].b) * ratio;
+        sp1CC = this->colorStart[1].a + ((f32)this->colorEnd[1].a - (f32)this->colorStart[1].a) * ratio;
+        sp1CB = this->colorStart[2].r + ((f32)this->colorEnd[2].r - (f32)this->colorStart[2].r) * ratio;
+        sp1CA = this->colorStart[2].g + ((f32)this->colorEnd[2].g - (f32)this->colorStart[2].g) * ratio;
+        sp1C9 = this->colorStart[2].b + ((f32)this->colorEnd[2].b - (f32)this->colorStart[2].b) * ratio;
+        sp1C8 = this->colorStart[2].a + ((f32)this->colorEnd[2].a - (f32)this->colorStart[2].a) * ratio;
+        sp1C7 = this->colorStart[3].r + ((f32)this->colorEnd[3].r - (f32)this->colorStart[3].r) * ratio;
+        sp1C6 = this->colorStart[3].g + ((f32)this->colorEnd[3].g - (f32)this->colorStart[3].g) * ratio;
+        sp1C5 = this->colorStart[3].b + ((f32)this->colorEnd[3].b - (f32)this->colorStart[3].b) * ratio;
+        sp1C4 = this->colorStart[3].a + ((f32)this->colorEnd[3].a - (f32)this->colorStart[3].a) * ratio;
```
I'm not going to type it all out, but this whole section is `F32_LERPIMP`

engineer124 commented on src/code/z_eff_tire_mark.c:
```diff
@@ -1,17 +1,287 @@
+    Vec3s* vtxList = colCtx->colHeader->vtxList;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eff_tire_mark/func_800AEF44.s")
+    if ((arg6 != 50) || (this->numElements >= 63) || (colPoly == NULL)) {
```
```suggestion
    if ((arg6 != 50) || (this->numElements >= ARRAY_COUNT(this->elements) - 1) || (colPoly == NULL)) {
```

engineer124 commented on src/code/z_eff_tire_mark.c:
```diff
@@ -1,17 +1,287 @@
+    EffectTireMarkInit* initParams = (EffectTireMarkInit*)initParamsx;
+    s32 i;
+
+    for (i = 0; i < 64; i++) {
```
```suggestion
    for (i = 0; i < ARRAY_COUNT(this->elements); i++) {
```

engineer124 commented on src/code/z_eff_tire_mark.c:
```diff
@@ -1,17 +1,287 @@
+    }
+
+    if (this->elements[0].life <= 0) {
+        for (j = 0; j < 63; j++) {
```
```suggestion
        for (j = 0; j < ARRAY_COUNT(this->elements) - 1; j++) {
```

engineer124 reviewed (commented):
Great work on the effects! As an fyi, I also brought in the effectsType enum in this PR: https://github.com/zeldaret/mm/pull/275, but the naming and where they went in this PR is exactly consistent with that, so no conflicts there. I'll remove the effectsType enum from that PR to save one of us git conflicts. A few comments:

Kenix3 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    EffectBlureElement* elem = &this->elements[index];
+
+    switch (this->calcMode) {
+        case 1:
```
Sounds good

Kenix3 commented on src/code/z_eff_blure.c:
```diff
@@ -66,40 +68,977 @@ void func_800A81F0(EffectBlure* this, Vec3f* p1, Vec3f* p2) {
+    CLOSE_DISPS(gfxCtx);
+}
+
+Vtx_t D_801AE26C[] = {
+    VTX_T(0, 0, 0, 0, 1024, 255, 255, 255, 255),
+    VTX_T(0, 0, 0, 0, 0, 255, 255, 255, 255),
+    VTX_T(0, 0, 0, 2048, 1024, 255, 255, 255, 255),
+    VTX_T(0, 0, 0, 2048, 0, 255, 255, 255, 255),
+};
+
+Vtx_t D_801AE2AC[] = {
+    VTX_T(0, 0, 0, 2048, 1024, 255, 255, 255, 255),
+    VTX_T(0, 0, 0, 2048, 0, 255, 255, 255, 255),
+    VTX_T(0, 0, 0, 2048, 1024, 255, 255, 255, 255),
+    VTX_T(0, 0, 0, 2048, 0, 255, 255, 255, 255),
+};
```
Contributors have voted that even trivial displayed vertices should be extracted.

Roman971 commented:
I updated the branch and changed the square vertices in `z_eff_shield_particle.c` to be extracted as an asset, tho I'm not sure if this is the right way to do it but I guess it works. I also added a few effect functions in `actorfixer.py`, in case there are any WIP actors using them.

hensldm reviewed (approved):
LGTM. The only outstanding issue I see left is the possible Vertices extraction in `z_eff_blure.c`. I don't have a strong opinion either way, so am happy to approve as is.

Kenix3 reviewed (changes requested):
Just waiting on extraction on the blure vertices. I know it's small, but the contributors voted no matter how trivial it should be extracted. Ready for merge after that.

Roman971 commented:
The `z_eff_shield_particle.c` vertices are extracted now (mentionned it in a previous comment), which were the ones we weren't sure about since they make a simple square. So the PR should be good to go in that regard, assuming I did the extraction right.

And to be clear, the `z_eff_blure.c` vertices aren't extracted since they don't really represent anything in their data form as discussed previously (I believe the terms used in the recent definition for extracted vertices were "points that constitute geometry primitives", which isn't the case here afaik)

Kenix3 commented:
> The `z_eff_shield_particle.c` vertices are extracted now (mentionned it in a previous comment), which were the ones we weren't sure about since they make a simple square. So the PR should be good to go in that regard, assuming I did the extraction right.
> 
> And to be clear, the `z_eff_blure.c` vertices aren't extracted since they don't really represent anything in their data form as discussed previously (I believe the terms used in the recent definition for extracted vertices were "points that constitute geometry primitives", which isn't the case here afaik)

Ah thankyou for reminding me Roman. Those were the points that the initial data just isn't used. That's fine.
